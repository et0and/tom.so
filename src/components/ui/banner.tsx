"use client"

import { SVGProps, useState } from "react"
import { Button } from "@/components/ui/button"

export default function Component() {
  const [showBanner, setShowBanner] = useState(true)
  const handleClose = () => {
    setShowBanner(false)
  }
  return (
    <>
      {showBanner && (
        <div className="bg-black text-white p-4 relative rounded-md m-4">
          <div className="flex items-center justify-between">
            <div className="flex items-center gap-2">
              <InfoIcon className="h-5 w-5" />
              <h1 className="text-lg font-semibold">Website upgrade</h1>
            </div>
            <Button
              variant="ghost"
              size="icon"
              className="absolute top-2 right-2 text-white hover:bg-gray-800 focus:bg-gray-800"
              onClick={handleClose}
            >
              <XIcon className="h-5 w-5" />
            </Button>
          </div>
          <p className="text-lg mt-4">This website is running on a new stack, and may contain a few bugs until these can be fixed. <a href="https://github.com/et0and/tom.so" className="underline hover:text-gray-300">More information</a>.</p>
        </div>
      )}
    </>
  )
}

function InfoIcon(props: JSX.IntrinsicAttributes & SVGProps<SVGSVGElement>) {
  return (
    <svg
      {...props}
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <circle cx="12" cy="12" r="10" />
      <path d="M12 16v-4" />
      <path d="M12 8h.01" />
    </svg>
  )
}


function XIcon(props: JSX.IntrinsicAttributes & SVGProps<SVGSVGElement>) {
  return (
    <svg
      {...props}
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <path d="M18 6 6 18" />
      <path d="m6 6 12 12" />
    </svg>
  )
}
