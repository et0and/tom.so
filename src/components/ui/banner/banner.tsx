import React from "react";
import classNames from "classnames";
import { AlertCircle } from "lucide-react";

export interface BannerProps {
  title: string;
  message: string;
  variant: "info" | "maintenance" | "error";
  showIcon?: boolean;
  linkName?: string;
  linkUrl?: string;
}

const variantStyles = {
  info: "bg-white text-black border-gray-200 dark:bg-stone-700 dark:text-white",
  maintenance: "bg-white text-black border-gray-200",
  error:
    "bg-black text-white border-black dark:bg-black dark:text-white dark:border-white",
};

export const Banner = ({
  title,
  message,
  variant,
  showIcon = true,
  linkName,
  linkUrl,
}: BannerProps) => {
  const outer = "leading-none relative my-4 w-full rounded-lg border px-4 py-4";
  const inner = "max-w-3xl w-full m-auto";

  const getIcon = () => {
    switch (variant) {
      case "info":
        return <AlertCircle size={20} />;
      case "maintenance":
      case "error":
        return <AlertCircle size={20} color="red" />;
      default:
        return null;
    }
  };

  return (
    <div
      role="region"
      aria-labelledby="tom-hackshaw-banner"
      className={classNames(outer, variantStyles[variant])}
    >
      <div className={classNames(inner, "flex mb-2")}>
        {showIcon && <span className="mr-2">{getIcon()}</span>}
        <h1 className="font-medium text-[16px]">{title}</h1>
      </div>
      <div className="leading-5">{message}</div>
      {linkName && linkUrl && (
        <a
          href={linkUrl}
          className="inline-flex no-underline p-4 mt-4 text-white transition rounded-full bg-blue-700 hover:bg-blue-900 hover:text-white hover:cursor-pointer"
        >
          {linkName}
        </a>
      )}
    </div>
  );
};
