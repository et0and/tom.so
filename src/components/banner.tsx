"use client"

import { SVGProps, useState } from "react"
import { Button } from "@/components/ui/button"

export default function Component() {
  const [showBanner, setShowBanner] = useState(true)
  const handleClose = () => {
    setShowBanner(false)
  }
  return (
    <>
      {showBanner && (
        <div className="bg-black text-white p-4 relative">
          <div className="flex items-center justify-between">
            <div className="flex items-center gap-2">
              <InfoIcon className="h-5 w-5" />
              <h3 className="text-lg font-bold">Important information</h3>
            </div>
            <Button
              variant="ghost"
              size="icon"
              className="absolute top-2 right-2 text-white hover:bg-gray-800 focus:bg-gray-800"
              onClick={handleClose}
            >
              <XIcon className="h-5 w-5" />
            </Button>
          </div>
          <p className="text-sm mt-4">Please review the latest updates to our terms of service.</p>
        </div>
      )}
    </>
  )
}

function InfoIcon(props: JSX.IntrinsicAttributes & SVGProps<SVGSVGElement>) {
  return (
    <svg
      {...props}
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <circle cx="12" cy="12" r="10" />
      <path d="M12 16v-4" />
      <path d="M12 8h.01" />
    </svg>
  )
}


function XIcon(props: JSX.IntrinsicAttributes & SVGProps<SVGSVGElement>) {
  return (
    <svg
      {...props}
      xmlns="http://www.w3.org/2000/svg"
      width="24"
      height="24"
      viewBox="0 0 24 24"
      fill="none"
      stroke="currentColor"
      strokeWidth="2"
      strokeLinecap="round"
      strokeLinejoin="round"
    >
      <path d="M18 6 6 18" />
      <path d="m6 6 12 12" />
    </svg>
  )
}
