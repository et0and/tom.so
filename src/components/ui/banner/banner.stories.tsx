import * as React from "react";
import { Banner } from "./banner";
import { StoryFn, Meta } from "@storybook/react";

// Define the metadata for the component
export default {
  title: "Components/Banner",
  component: Banner,
  argTypes: {
    variant: {
      control: "select",
      options: ["info", "maintenance", "error", "link"],
      description: "The visual style of the banner",
    },
  },
} as Meta<typeof Banner>;

// Create a template for the component
const Template: StoryFn<typeof Banner> = (args) => <Banner {...args} />;

// Create different stories for each variant and size
export const Default = Template.bind({});
Default.args = {
  title: "Default banner",
  message: "This is a default banner message.",
  variant: "info",
  showIcon: true,
};

export const Maintenance = Template.bind({});
Maintenance.args = {
  title: "Maintenance banner",
  message: "This is a maintenance banner message.",
  variant: "maintenance",
  showIcon: true,
};

export const Error = Template.bind({});
Error.args = {
  title: "Error banner",
  message: "This is a error banner message.",
  variant: "error",
  showIcon: true,
};

export const WithoutIcon = Template.bind({});
WithoutIcon.args = {
  title: "Banner without Icon",
  message: "This is a banner without an icon.",
  variant: "info",
  showIcon: false,
};
