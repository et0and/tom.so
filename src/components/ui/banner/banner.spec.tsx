import React from "react";
import { render } from "@testing-library/react";
import { Banner, BannerProps } from "./banner";

// Mock the Lucide React component
jest.mock("lucide-react", () => ({
  AlertCircle: () => <div data-testid="mock-alert-circle" />,
}));

describe("<Banner/>", () => {
  const defaultProps: BannerProps = {
    title: "Test Title",
    message: "Test Message",
    variant: "info",
  };

  it("renders info banner correctly", () => {
    const { asFragment } = render(<Banner {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders maintenance banner correctly", () => {
    const { asFragment } = render(
      <Banner {...defaultProps} variant="maintenance" />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders error banner correctly", () => {
    const { asFragment } = render(<Banner {...defaultProps} variant="error" />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders banner without icon", () => {
    const { asFragment } = render(
      <Banner {...defaultProps} showIcon={false} />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders banner with link", () => {
    const { asFragment } = render(
      <Banner
        {...defaultProps}
        linkName="Click here"
        linkUrl="https://example.com"
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
