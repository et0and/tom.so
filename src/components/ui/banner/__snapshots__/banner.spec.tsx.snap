// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Banner/> renders banner with link 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tom-hackshaw-banner"
    class="leading-none relative my-4 w-full rounded-lg border px-4 py-4 bg-white text-black border-gray-200 dark:bg-stone-700 dark:text-white"
    role="region"
  >
    <div
      class="max-w-3xl w-full m-auto flex mb-2"
    >
      <span
        class="mr-2"
      >
        <div
          data-testid="mock-alert-circle"
        />
      </span>
      <h1
        class="font-medium text-[16px]"
      >
        Test Title
      </h1>
    </div>
    <div
      class="leading-5"
    >
      Test Message
    </div>
    <a
      class="inline-flex no-underline p-4 mt-4 text-white transition rounded-full bg-blue-700 hover:bg-blue-900 hover:text-white hover:cursor-pointer"
      href="https://example.com"
    >
      Click here
    </a>
  </div>
</DocumentFragment>
`;

exports[`<Banner/> renders banner without icon 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tom-hackshaw-banner"
    class="leading-none relative my-4 w-full rounded-lg border px-4 py-4 bg-white text-black border-gray-200 dark:bg-stone-700 dark:text-white"
    role="region"
  >
    <div
      class="max-w-3xl w-full m-auto flex mb-2"
    >
      <h1
        class="font-medium text-[16px]"
      >
        Test Title
      </h1>
    </div>
    <div
      class="leading-5"
    >
      Test Message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Banner/> renders error banner correctly 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tom-hackshaw-banner"
    class="leading-none relative my-4 w-full rounded-lg border px-4 py-4 bg-black text-white border-black dark:bg-black dark:text-white dark:border-white"
    role="region"
  >
    <div
      class="max-w-3xl w-full m-auto flex mb-2"
    >
      <span
        class="mr-2"
      >
        <div
          data-testid="mock-alert-circle"
        />
      </span>
      <h1
        class="font-medium text-[16px]"
      >
        Test Title
      </h1>
    </div>
    <div
      class="leading-5"
    >
      Test Message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Banner/> renders info banner correctly 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tom-hackshaw-banner"
    class="leading-none relative my-4 w-full rounded-lg border px-4 py-4 bg-white text-black border-gray-200 dark:bg-stone-700 dark:text-white"
    role="region"
  >
    <div
      class="max-w-3xl w-full m-auto flex mb-2"
    >
      <span
        class="mr-2"
      >
        <div
          data-testid="mock-alert-circle"
        />
      </span>
      <h1
        class="font-medium text-[16px]"
      >
        Test Title
      </h1>
    </div>
    <div
      class="leading-5"
    >
      Test Message
    </div>
  </div>
</DocumentFragment>
`;

exports[`<Banner/> renders maintenance banner correctly 1`] = `
<DocumentFragment>
  <div
    aria-labelledby="tom-hackshaw-banner"
    class="leading-none relative my-4 w-full rounded-lg border px-4 py-4 bg-white text-black border-gray-200"
    role="region"
  >
    <div
      class="max-w-3xl w-full m-auto flex mb-2"
    >
      <span
        class="mr-2"
      >
        <div
          data-testid="mock-alert-circle"
        />
      </span>
      <h1
        class="font-medium text-[16px]"
      >
        Test Title
      </h1>
    </div>
    <div
      class="leading-5"
    >
      Test Message
    </div>
  </div>
</DocumentFragment>
`;
